framework/math/ParallelVector/ParallelSTLVector.h

```diff
@@ -0,0 +1,158 @@
+#ifndef PARALLEL_VECTOR_H
+#define PARALLEL_VECTOR_H
+
+#include "ParallelVector.h"
+#include "math/chi_math.h"
+
+#include <vector>
+#include <cstdint>
+#include <map>
+
+#include <mpi.h>
+
+namespace chi_math
+{
+
+/**
+ * An implementation of a parallel vector.
+ */
+class ParallelSTLVector : public ParallelVector
+{
+public:
+  /**
+   * Initialize a parallel vector with the given local and global sizes with
+   * the given communicator whose entries are set to zero.
+   */
+  ParallelSTLVector(uint64_t local_size,
+                    uint64_t global_size,
+                    MPI_Comm communicator);
+
+  /**Copy constructor.*/
+  ParallelSTLVector(const ParallelSTLVector& other);
+
+  /**Move constructor.*/
+  ParallelSTLVector(ParallelSTLVector&& other) noexcept;
+
+  std::unique_ptr<ParallelVector> MakeCopy() const override;
+  std::unique_ptr<ParallelVector> MakeNewVector() const override;
```

Comment:
Oh damn. Base class missing documentation. Apologies.

```c++
/**Creates a copy of the vector datastructures AND values.
  * This routine requires no communication.*/
  virtual std::unique_ptr<ParallelVector> MakeCopy() const = 0;

  /**Creates a copy of the vector datastructures but NOT the values. The
  * values are defaulted to zero. This routine requires no communication.*/
  virtual std::unique_ptr<ParallelVector> MakeNewVector() const = 0;
```

Reply:
This is akin to PETSc's `VecCopy` and `VecClone` and they are very useful. Consider the code below:
```
r_t_tp1_ = r.MakeNewVector();
SetModeToTimeOnly();
eq_system_->ComputeResidual(x, *r_t_tp1_);

if (TimeIDListHasID(r_time_ids, TimeID::T_PLUS_1))
{
    r_x_tp1_ = r.MakeNewVector();
    SetModeToNonTimeOnly();
    eq_system_->ComputeResidual(x, *r_x_tp1_);
    r_map_[TimeID::T_PLUS_1] = (&(*r_x_tp1_));
}
```

Reply:
This is what a MakeCopy, like you suggest will look like:
```
r_t_tp1_ = r.MakeCopy();
r_t_tp1.Set(0.0);

SetModeToTimeOnly();
eq_system_->ComputeResidual(x, *r_t_tp1_);

if (TimeIDListHasID(r_time_ids, TimeID::T_PLUS_1))
{
    r_x_tp1_ = r.MakeNewVector();
    r_x_tp1_.Set(0.0);
    SetModeToNonTimeOnly();
    eq_system_->ComputeResidual(x, *r_x_tp1_);
    r_map_[TimeID::T_PLUS_1] = (&(*r_x_tp1_));
}
```

Reply:
Another example:
```
std::vector<std::unique_ptr<ParallelVector>>
EquationSystem::InitSolutionHistory()
{
  std::vector<std::unique_ptr<ParallelVector>> old_solution_vectors;
  for (size_t t = 0; t < num_solution_histories_; ++t)
  {
    auto vec = main_solution_vector_->MakeNewVector();
    old_solution_vectors.push_back(std::move(vec));
  }

  return old_solution_vectors;
}
```

Reply:
I'm still not sure the purpose of returning pointers instead of objects here.

Reply:
From the Slack discussion:
The problem I encountered is that when are in a routine where you passed the base ParallelVector by reference, you can't copy it. Its an abstract class. There is no copy constructor defined and you cannot make a virtual one because ParallelSTL vector or GhostedParallelSTL vector doesn't know what its dealing with.

Petsc hides this very well from us because they're Vec  is actually a typedef on a pointer. Inherently we should always store our vectors as pointers for polymorphism

Now, why a unique_ptr. 1) The ownership is very clear. The vector that spawns a new vector should no longer hold on to it so it shouldn't be a smart_pointer. 2) A unique_ptr can easily be converted to a shared_ptr but a shared_ptr cannot be converted to a shared_ptr.